File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/levelOrder/199. Binary Tree Right Side View.py
from collections import deque
from typing import List
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        right_view = []
        queue = deque([root])
        while queue:
            level_size = len(queue)
            for i in range(level_size):
                node = queue.popleft()
                if i == level_size - 1:
                    right_view.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        return right_view

# Test case
root = TreeNode(1, TreeNode(2, None, TreeNode(5)), TreeNode(3, None, TreeNode(4)))
sol = Solution()
print(sol.rightSideView(root))  # [1, 3, 4]

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/levelOrder/199. Binary Tree Right Side View.cpp
// 199. Binary Tree Right Side View
// https://leetcode.com/problems/binary-tree-right-side-view/
// clang++ 199.\ Binary\ Tree\ Right\ Side\ View.cpp  -std=c++17

#include "../../../printUtils.hpp"
#include <vector>

template <typename T>
std::vector<T> rightSideView(TreeNode<T>* root) {
    std::vector<T> result;
    if (root == nullptr) return result;

    std::queue<TreeNode<T>*> q;
    q.push(root);

    while (!q.empty()) {
        int size = q.size();
        for (int i = 0; i < size; i++) {
            TreeNode<T>* curr = q.front();
            q.pop();

            if (i == size - 1) {
                result.push_back(curr->val);
            }

            if (curr->left) q.push(curr->left);
            if (curr->right) q.push(curr->right);
            }
        }
    return result;
}

int main() {
    TreeNode<int>* root = buildTree({1, 2, 3, -1, 5, -1, 4}, -1);
    std::cout << "Right Side View of Binary Tree:" << std::endl;
    printArray(rightSideView(root));
    deleteTree(root);
    
    return 0;
}
